File: clients/csharp/examples/collections-find-many-iterate-manual.cs
using System.Text.Json;
using DataStax.AstraDB.DataApi;
using DataStax.AstraDB.DataApi.Collections;
using DataStax.AstraDB.DataApi.Core;
using DataStax.AstraDB.DataApi.Core.Query;

namespace Examples;

public class Program
{
  static async Task Main()
  {
    // Get an existing collection
    var client = new DataAPIClient();
    var database = client.GetDatabase(
      "**API_ENDPOINT**",
      "**APPLICATION_TOKEN**"
    );
    var collection = database.GetCollection("**COLLECTION_NAME**");

    // Create the filter
    var filterBuilder = Builders<Document>.CollectionFilter;
    var filter = filterBuilder.And(
      filterBuilder.Eq("is_checked_out", false),
      filterBuilder.Lt("number_of_pages", 300)
    );

    // Get the first page
    var cursor1 = collection.Find(filter);
    var page1 = await cursor1.FetchNextPageAsync();
    var results1 = page1.Results;
    foreach (var document in results1)
    {
      Console.WriteLine(JsonSerializer.Serialize(document));
    }
    var paginationState1 = page1.NextPageState;

    // Get the next page
    if (paginationState1 != null)
    {
      var cursor2 = collection.Find(
        filter,
        new CollectionFindOptions<Document>()
        {
          InitialPageState = paginationState1,
        }
      );
      var page2 = await cursor2.FetchNextPageAsync();
      var results2 = page2.Results;
      foreach (var document in results2)
      {
        Console.WriteLine(JsonSerializer.Serialize(document));
      }
      var paginationState2 = page2.NextPageState;
    }
  }
}

File: clients/python/examples/collections-find-many-iterate-manual.py
from astrapy import DataAPIClient

# Get an existing collection
client = DataAPIClient()
database = client.get_database(
    "**API_ENDPOINT**", token="**APPLICATION_TOKEN**"
)
collection = database.get_collection("**COLLECTION_NAME**")

# Create the filter
filter = {
    "$and": [
        {"is_checked_out": False},
        {"number_of_pages": {"$lt": 300}},
    ]
}

# Get the first page
cursor_1 = collection.find(filter)
page_1 = cursor_1.fetch_next_page()
results_1 = page_1.results
for document in results_1:
    print(document)
pagination_state_1 = page_1.next_page_state

# Get the next page
if pagination_state_1:
    cursor_2 = collection.find(
        filter,
        initial_page_state=pagination_state_1,
    )
    page_2 = cursor_2.fetch_next_page()
    results_2 = page_2.results
    for document in results_2:
        print(document)
    pagination_state_2 = page_2.next_page_state
